espresso/environment/8_reorg_test.go

```diff
@@ -0,0 +1,115 @@
+package environment_test
+
+import (
+	"context"
+	"math/big"
+	"testing"
+	"time"
+
+	env "github.com/ethereum-optimism/optimism/espresso/environment"
+	"github.com/ethereum-optimism/optimism/op-e2e/system/e2esys"
+	"github.com/ethereum-optimism/optimism/op-e2e/system/helpers"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/stretchr/testify/require"
+)
+
+// TestBatcherWaitForFinality is a test that attempts to make sure that the batcher waits for the
+// derived L1 block to be finalized before submitting a new block.
+//
+// This tests is designed to evaluate Test 8.2 as outlined within the Espresso Celo Integration
+// plan. It has stated task definition as follows:
+//
+//	Arrange:
+//		Run the sequencer and the batcher in Espresso mode.
+//	Act:
+//		Send two transactions.
+//	Assert:
+//		The batcher doesn't submit the second transaction to the L1 immediatly.
+//		After the first block is finalized the L1, the batcher submits the second transaction.
+func TestBatcherWaitForFinality(t *testing.T) {
+	// Basic test setup.
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	launcher := new(env.EspressoDevNodeLauncherDocker)
+	system, espressoDevNode, err := launcher.StartDevNet(ctx, t, 0)
+	if have, want := err, error(nil); have != want {
+		t.Fatalf("failed to start dev environment with espresso dev node:\nhave:\n\t\"%v\"\nwant:\n\t\"%v\"\n", have, want)
+	}
+	defer env.Stop(t, system)
+	defer env.Stop(t, espressoDevNode)
+
+	caffNode, err := env.LaunchDecaffNode(t, system, espressoDevNode)
+	if have, want := err, error(nil); have != want {
+		t.Fatalf("failed to start caff node:\nhave:\n\t\"%v\"\nwant:\n\t\"%v\"\n", have, want)
+	}
+	defer env.Stop(t, caffNode)
+
+	l2Seq := system.NodeClient("sequencer")
+	l2Verif := system.NodeClient(e2esys.RoleVerif)
+	rollupClient := system.RollupClient("verifier")
+
+	// Record the initial sync status.
+	initialSeqStatus, err := rollupClient.SyncStatus(context.Background())
+	require.NoError(t, err)
+	initialFinalizedL1Number := initialSeqStatus.FinalizedL1.Number
+	initialCurrentL1Number := initialSeqStatus.CurrentL1.Number
+
+	// Send two transactions.
```

Comment:
Are these transactions expected to be included in the first / (resp.) second L2 block?

Reply:
Actually the test passes without sending these transactions. So what is the purpose of these transactions? Maybe they are not needed as batches are created automatically anyways.

Reply:
Good point! I'm going to remove these transactions. I'll also add one more check before waiting for the finality, to fully test the "no submission before L1 origin finality".